swayrun/main.c

```diff
@@ -0,0 +1,71 @@
+#define _POSIX_C_SOURCE 200809L // for getline
+#include <errno.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include "stringop.h"
+
+#define SWAY_COMMAND "sway"
+
+char allowed_shell(char *shell) {
+	FILE *fp;
+	char *line = NULL;
+	size_t len = 0;
+	char allowed = false;
+
+	if (strstr(shell, "false") != NULL || strstr(shell, "nologin") != NULL) {
+		return false;
+	}
+
+	fp = fopen("/etc/shells", "r");
+	if (fp == NULL) {
+		return true;
+	}
+
+	while (getline(&line, &len, fp) != -1) {
+		strip_whitespace(line);
+		if (strcmp(shell, line) == 0) {
+			allowed = true;
+			break;
+		}
+	}
+
+	fclose(fp);
+	if (line) {
+		free(line);
+	}
+	return allowed;
+}
+
+int main(int argc, char **argv) {
+	char *shell = getenv("SHELL");
+
+	if (shell && allowed_shell(shell)) {
+		// 3 exec arguments + argc + argv[argc] NULL pointer
+		int exec_argc = 4 + argc;
+		char **exec_argv = malloc(exec_argc * sizeof(char*));
+
+		// Prefix - to shell path to indicate login shell
+		char *login_shell = malloc(strlen(shell) + 2);
+		strcpy(login_shell, "-");
+		strcat(login_shell, shell);
+
+		// Build the argumrnts to exec
+		memcpy(exec_argv + 3, argv, (argc + 1) * sizeof(argv));
+		exec_argv[0] = login_shell;
+		exec_argv[1] = "-c";
+		exec_argv[2] = "exec " SWAY_COMMAND " \"$@\"";
+		exec_argv[3] = shell;
+
+		execvp(shell, exec_argv);
+		fprintf(stderr, "Could not run %s using login shell: %s\n", SWAY_COMMAND, shell);
+	} else {
+		argv[0] = SWAY_COMMAND;
+		execvp(SWAY_COMMAND, argv);
+		fprintf(stderr, "Could not run %s\n", SWAY_COMMAND);
+	}
```

Comment:
Why do we care if they want to run it with some shell which isn't in /etc/shells?

Reply:
I modeled the approach from the `gnome-session` wrapper. It checks `/etc/shells` and also does not accept nologin or false. I suppose it is probably not necessary since by the time this gets called, the DM has already decided to allow or deny a login. 

Would you prefer the check be removed?

Reply:
I don't particularly care either way, just seems unnecessary. I'm not -1 or +1 on this

Reply:
I'd prefer to drop it, it simplifies the code.

Reply:
@emersion I agree it would simplify things. Can you see any security concerns with using an “unapproved” login shell? `chsh` and friends confirm a shell is in `/etc/shells` before running.

Reply:
Yeah, but we assume the environment from which we're started is safe. If it's not the case, much more nasty things could be done, e.g. via `LD_PRELOAD` or `PATH`.

Reply:
Sounds reasonable. The `allowed_shell` check has been removed.